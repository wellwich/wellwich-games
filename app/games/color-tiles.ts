import Phaser from 'phaser';

const GAME_CONFIG = {
	DEFAULT_WIDTH: 640,
	DEFAULT_HEIGHT: 480,
	DEFAULT_FONT: {
		fontFamily: 'The Strong Gamer',
		fontSize: '128px',
		color: '#ffffff',
	} as Phaser.Types.GameObjects.Text.TextStyle,
};

class TitleScene extends Phaser.Scene {
	constructor() {
		super('titleScene');
	}
	create() {
		const background = this.add.rectangle(
			0,
			0,
			GAME_CONFIG.DEFAULT_WIDTH,
			GAME_CONFIG.DEFAULT_HEIGHT,
			0x444444,
			0.5,
		);
		background.setOrigin(0, 0);
		background.setInteractive();
		const fullScreen = this.add
			.text(GAME_CONFIG.DEFAULT_WIDTH - 16, 16, 'Full Screen', {
				fontSize: '16px',
				fontFamily: '"font"',
			})
			.setOrigin(1, 0);
		fullScreen.setInteractive();
		fullScreen.on('pointerup', () => {
			if (this.scale.isFullscreen) {
				this.scale.stopFullscreen();
			} else {
				this.scale.startFullscreen();
			}
		});
		const title = this.add
			.text(
				GAME_CONFIG.DEFAULT_WIDTH / 2,
				GAME_CONFIG.DEFAULT_HEIGHT / 2 - 50,
				'カラータイル',
				{ fontSize: '64px', color: '#fff', fontFamily: 'monospace' },
			)
			.setOrigin(0.5, 0.5);
		const start = this.add
			.text(
				GAME_CONFIG.DEFAULT_WIDTH / 2,
				GAME_CONFIG.DEFAULT_HEIGHT / 2 + 100,
				'Tap to start',
				{ fontSize: '16px', fontFamily: '"font"' },
			)
			.setOrigin(0.5, 0.5);
		// 画面のどこかをタップしたらゲーム画面へ
		background.on('pointerup', () => {
			this.scene.start('gameScene');
		});
		if (window.matchMedia?.('(max-device-width: 640px)').matches) {
			const background = this.add.rectangle(
				0,
				0,
				GAME_CONFIG.DEFAULT_WIDTH,
				GAME_CONFIG.DEFAULT_HEIGHT,
				0x444444,
			);
			background.setInteractive();
			background.setOrigin(0, 0);
			this.add
				.text(
					GAME_CONFIG.DEFAULT_WIDTH / 2,
					GAME_CONFIG.DEFAULT_HEIGHT / 2,
					'横画面にしてください',
					GAME_CONFIG.DEFAULT_FONT,
				)
				.setFontSize(32)
				.setOrigin(0.5);
			const fixScreen = this.add
				.text(
					GAME_CONFIG.DEFAULT_WIDTH / 2,
					GAME_CONFIG.DEFAULT_HEIGHT / 2 + 150,
					'横画面にしたらここをタップ',
					GAME_CONFIG.DEFAULT_FONT,
				)
				.setFontSize(32)
				.setOrigin(0.5);
			fixScreen.setInteractive();
			fixScreen.on('pointerup', () => {
				this.scene.start('titleScene');
			});
		}
	}
}

class GameScene extends Phaser.Scene {
	constructor() {
		super('gameScene');
	}

	private blankTileColor = 0x999999;
	private redTileColor = 0xff0000;
	private greenTileColor = 0x00ff00;
	private blueTileColor = 0x00c9e8;
	private yellowTileColor = 0xeeee00;
	private purpleTileColor = 0x800080;
	private pinkTileColor = 0xff44ff;
	private colorArray: number[] = [
		this.redTileColor,
		this.greenTileColor,
		this.blueTileColor,
		this.yellowTileColor,
		this.purpleTileColor,
		this.pinkTileColor,
		this.blankTileColor,
	];
	private tiles!: Phaser.GameObjects.Group;

	preload() {
		this.load.audio('miss', 'https://assets.wellwich.com/sounds/miss.mp3');
		this.load.audio(
			'correct',
			'https://assets.wellwich.com/sounds/correct.mp3',
		);
	}

	create() {
		// ズームレベルを2に設定
		this.cameras.main.setBackgroundColor(0x999999);

		let score = 0;
		const scoreText = this.add
			.text(8, 8, `${score}`, {
				fontSize: '32px',
				fontFamily: GAME_CONFIG.DEFAULT_FONT.fontFamily,
				color: '#ffffff',
			})
			.setDepth(1);

		const missSound = this.sound.add('miss');
		const correctSound = this.sound.add('correct');

		// 空白ありでタイルcolorArray内の色でマス目状にランダムに配置
		this.tiles = this.add.group();
		const tileSize = 26;
		const tileSpacing = 2;
		const tileWidth = tileSize + tileSpacing;
		const tileHeight = tileSize + tileSpacing;
		const rows = 13;
		const cols = 20;
		// タイル全体のサイズを計算
		const totalWidth = cols * tileWidth; // 最後のスペーシングは不要
		const totalHeight = rows * tileHeight; // 最後のスペーシングは不要
		// タイルグループの開始位置を計算
		const startX = (GAME_CONFIG.DEFAULT_WIDTH - totalWidth) / 2;
		const startY = (GAME_CONFIG.DEFAULT_HEIGHT - totalHeight) / 2 + 8;
		for (let i = 0; i < rows; i++) {
			for (let j = 0; j < cols; j++) {
				const tile = this.add.rectangle(
					startX + j * tileWidth,
					startY + i * tileHeight,
					tileSize,
					tileSize,
					Phaser.Math.RND.pick(this.colorArray),
				);
				tile.setData('color', tile.fillColor);
				tile.setData('row', i);
				tile.setData('col', j);
				tile.setInteractive();
				tile.on('pointerup', () => {
					// タップされた空白のタイルの行と列を取得
					if (tile.getData('color') === this.blankTileColor) {
						const row = tile.getData('row');
						const col = tile.getData('col');
						// 上下左右のタイルを取得
						const tileUp = this.getTileToMove(row, col, -1, 0);
						const tileDown = this.getTileToMove(row, col, 1, 0);
						const tileLeft = this.getTileToMove(row, col, 0, -1);
						const tileRight = this.getTileToMove(row, col, 0, 1);
						// 上下左右の組み合わせ全パターンをチェックして、同じ色のタイルがあれば空白のタイルと入れ替え
						const deletableTiles: Phaser.GameObjects.Rectangle[] = [];
						if (tileUp) {
							if (tileDown) {
								if (tileUp.getData('color') === tileDown.getData('color')) {
									deletableTiles.push(tileUp, tileDown);
								}
							}
							if (tileLeft) {
								if (tileUp.getData('color') === tileLeft.getData('color')) {
									deletableTiles.push(tileUp, tileLeft);
								}
							}
							if (tileRight) {
								if (tileUp.getData('color') === tileRight.getData('color')) {
									deletableTiles.push(tileUp, tileRight);
								}
							}
						}
						if (tileDown) {
							if (tileLeft) {
								if (tileDown.getData('color') === tileLeft.getData('color')) {
									deletableTiles.push(tileDown, tileLeft);
								}
							}
							if (tileRight) {
								if (tileDown.getData('color') === tileRight.getData('color')) {
									deletableTiles.push(tileDown, tileRight);
								}
							}
						}
						if (tileLeft && tileRight) {
							if (tileLeft.getData('color') === tileRight.getData('color')) {
								deletableTiles.push(tileLeft, tileRight);
							}
						}
						if (deletableTiles.length > 0) {
							for (const t of deletableTiles) {
								t.setData('color', this.blankTileColor);
								t.fillColor = this.blankTileColor;
							}
						}
						if (deletableTiles.length > 0) {
							score += 100;
							scoreText.setText(`${score}`);
							correctSound.play();
						} else {
							score -= 100;
							scoreText.setText(`${score}`);
							missSound.play();
						}
					}
				});
				this.tiles.add(tile);
			}
		}
		this.tiles.setOrigin(0, 0);
	}

	// 空白のみ進める直線状で最初にぶつかる空白じゃないタイルを取得
	private getTileToMove(
		row: number,
		col: number,
		rowDir: number,
		colDir: number,
	): Phaser.GameObjects.Rectangle | null {
		let currentRow = row + rowDir;
		let currentCol = col + colDir;
		while (
			currentRow >= 0 &&
			currentRow < 13 &&
			currentCol >= 0 &&
			currentCol < 20
		) {
			const tile = this.tiles
				.getChildren()
				.find(
					(t: Phaser.GameObjects.GameObject) =>
						t.getData('row') === currentRow && t.getData('col') === currentCol,
				) as Phaser.GameObjects.Rectangle;
			if (tile.getData('color') !== this.blankTileColor) {
				return tile;
			}
			currentRow += rowDir;
			currentCol += colDir;
		}
		return null;
	}
}

const config: Phaser.Types.Core.GameConfig = {
	type: Phaser.AUTO,
	title: 'カラータイル',
	width: GAME_CONFIG.DEFAULT_WIDTH,
	height: GAME_CONFIG.DEFAULT_HEIGHT,
	scene: [TitleScene, GameScene],
	render: {
		pixelArt: true,
	},
	scale: {
		parent: 'phaser-game',
		mode: Phaser.Scale.FIT,
		fullscreenTarget: 'phaser-game',
		autoCenter: Phaser.Scale.CENTER_HORIZONTALLY,
		zoom: Phaser.Scale.MAX_ZOOM,
	},
};

export default function initPhaserGame(parent: HTMLElement) {
	if (config.scale) {
		config.scale.parent = parent;
		config.scale.fullscreenTarget = parent;
	}
	return new Phaser.Game(config);
}
