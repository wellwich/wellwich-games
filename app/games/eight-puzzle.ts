// src/phaserGame.ts
import Phaser from 'phaser';

const GAME_CONFIG = {
	DEFAULT_WIDTH: 1280,
	DEFAULT_HEIGHT: 960,
	DEFAULT_FONT: {
		fontFamily: 'The Strong Gamer',
		fontSize: '128px',
		color: '#ffffff',
	} as Phaser.Types.GameObjects.Text.TextStyle,
	MONO_FONT: {
		fontFamily: 'monospace',
		fontSize: '64px',
		color: '#ffffff',
	} as Phaser.Types.GameObjects.Text.TextStyle,
	CARD_FONT: {
		fontFamily: 'monospace',
		fontSize: '48px',
		color: '#000000',
	} as Phaser.Types.GameObjects.Text.TextStyle,
};

class TitleScene extends Phaser.Scene {
	constructor() {
		super('titleScene');
	}
	create() {
		const background = this.add.rectangle(
			0,
			0,
			GAME_CONFIG.DEFAULT_WIDTH,
			GAME_CONFIG.DEFAULT_HEIGHT,
			0x444444,
			0.5,
		);
		background.setOrigin(0, 0);
		background.setInteractive();
		const fullScreen = this.add
			.text(GAME_CONFIG.DEFAULT_WIDTH - 16, 16, 'Full Screen', {
				fontSize: '32px',
				fontFamily: 'The Strong Gamer',
			})
			.setOrigin(1, 0);
		fullScreen.setInteractive();
		fullScreen.on('pointerup', () => {
			if (this.scale.isFullscreen) {
				this.scale.stopFullscreen();
			} else {
				this.scale.startFullscreen();
			}
		});
		const title = this.add
			.text(
				GAME_CONFIG.DEFAULT_WIDTH / 2,
				GAME_CONFIG.DEFAULT_HEIGHT / 2 - 50,
				'8パズル',
				{ fontSize: '96px', fontFamily: 'monospace' },
			)
			.setOrigin(0.5, 0.5);
		const start = this.add
			.text(
				GAME_CONFIG.DEFAULT_WIDTH / 2,
				GAME_CONFIG.DEFAULT_HEIGHT / 2 + 100,
				'Tap to start',
				{ fontSize: '64px', fontFamily: 'The Strong Gamer' },
			)
			.setOrigin(0.5, 0.5)
			.setFontSize(32);
		// 画面のどこかをタップしたらゲーム画面へ
		background.on('pointerup', () => {
			this.scene.start('gameScene', {
				level: 1,
				visibleTime: 3000,
				missCount: 0,
			});
		});
		if (window.matchMedia?.('(max-device-width: 640px)').matches) {
			const background = this.add.rectangle(
				0,
				0,
				GAME_CONFIG.DEFAULT_WIDTH,
				GAME_CONFIG.DEFAULT_HEIGHT,
				0x444444,
			);
			background.setInteractive();
			background.setOrigin(0, 0);
			this.add
				.text(
					GAME_CONFIG.DEFAULT_WIDTH / 2,
					GAME_CONFIG.DEFAULT_HEIGHT / 2,
					'横画面にしてください',
					GAME_CONFIG.DEFAULT_FONT,
				)
				.setFontSize(32)
				.setOrigin(0.5);
			const fixScreen = this.add
				.text(
					GAME_CONFIG.DEFAULT_WIDTH / 2,
					GAME_CONFIG.DEFAULT_HEIGHT / 2 + 150,
					'横画面にしたらここをタップ',
					GAME_CONFIG.DEFAULT_FONT,
				)
				.setFontSize(32)
				.setOrigin(0.5);
			fixScreen.setInteractive();
			fixScreen.on('pointerup', () => {
				this.scale.startFullscreen();
				this.scene.start('titleScene');
			});
		}
	}
}

class GameScene extends Phaser.Scene {
	private tiles!: (Phaser.GameObjects.Image | null)[][];
	private correctTiles!: (Phaser.GameObjects.Image | null)[][];

	private gameWidth = GAME_CONFIG.DEFAULT_WIDTH;
	private gameHeight = GAME_CONFIG.DEFAULT_HEIGHT;

	private clearCount = 0;
	private clearCountText!: Phaser.GameObjects.Text;

	private timer = 60;
	private timeText!: Phaser.GameObjects.Text;

	private isShuffling = false;

	init(data: { clearCount: number; timer: number }) {
		if (data.clearCount) {
			this.clearCount = data.clearCount || 0;
		}
		if (data.timer) {
			this.timer = data.timer || 60;
		}
	}

	constructor() {
		super({ key: 'gameScene' });
	}

	preload() {
		const graphics = this.add.graphics();
		graphics.fillStyle(0x6666ff);
		graphics.fillRect(0, 0, 200, 200);
		graphics.lineStyle(4, 0x000000);
		graphics.strokeRect(0, 0, 200, 200);

		for (let i = 1; i <= 8; i++) {
			if (this.textures.exists(`tile${i}`)) {
				continue;
			}

			const number = this.add
				.text(100, 90, i.toString(), {
					color: '#000000',
					fontFamily: GAME_CONFIG.DEFAULT_FONT.fontFamily,
					fontSize: 100,
				})
				.setOrigin(0.5);
			const renderTexture = this.add.renderTexture(0, 0, 200, 200);
			renderTexture.draw(graphics);
			renderTexture.draw(number);
			renderTexture.saveTexture(`tile${i}`);
			renderTexture.destroy();
			number.destroy();
		}

		graphics.destroy();
	}

	create() {
		this.timeText = this.add.text(10, 10, `Time: ${this.timer}`, {
			color: '#ffffff',
			fontFamily: GAME_CONFIG.DEFAULT_FONT.fontFamily,
			fontSize: 30,
		});
		this.timeText.setOrigin(0, 0);
		this.timeText.setScrollFactor(0);
		this.timeText.setDepth(1);
		this.timeText.setInteractive();
		this.timeText.on('pointerdown', () => {
			this.scene.start('gameScene', { clearCount: 0, timer: 60 });
		});

		this.clearCountText = this.add.text(10, 50, `Clear: ${this.clearCount}`, {
			color: '#ffffff',
			fontFamily: GAME_CONFIG.DEFAULT_FONT.fontFamily,
			fontSize: 30,
		});
		this.clearCountText.setOrigin(0, 0);
		this.clearCountText.setScrollFactor(0);
		this.clearCountText.setDepth(1);

		this.time.addEvent({
			delay: 1000,
			callback: () => {
				this.timer--;
				this.timeText.setText(`Time: ${this.timer}`);
				if (this.timer <= 0) {
					this.clearCount = 0; // 追加
					this.scene.start('gameScene', {
						clearCount: this.clearCount,
						timer: 60,
					});
				}
			},
			loop: true,
		});

		const tiles: (Phaser.GameObjects.Image | null)[][] = [];
		const correctTiles: Array<Array<Phaser.GameObjects.Image | null>> = [];
		const offsetX = this.gameWidth / 2 - 200;
		const offsetY = this.gameHeight / 2 - 200;
		for (let i = 0; i < 3; i++) {
			tiles[i] = [];
			correctTiles[i] = [];
			for (let j = 0; j < 3; j++) {
				if (i === 2 && j === 2) {
					tiles[i].push(null);
					correctTiles[i].push(null);
				} else {
					const tile = this.add
						.image(offsetX + 200 * i, offsetY + 200 * j, `tile${i + j * 3 + 1}`)
						.setInteractive()
						.setOrigin(0.5);
					tiles[i].push(tile);
					correctTiles[i].push(tile);
				}
			}
		}
		tiles[2][2] = null;
		correctTiles[2][2] = null;
		this.tiles = tiles;
		this.correctTiles = correctTiles;
		this.shuffleTiles();
		this.input.on(
			'gameobjectdown',
			(pointer: Phaser.Input.Pointer, gameObject: Phaser.GameObjects.Image) => {
				const tile = gameObject;
				const x = Math.floor((tile.x - offsetX) / 200);
				const y = Math.floor((tile.y - offsetY) / 200);
				if (this.canMoveTile(x, y)) {
					this.moveTile(x, y);
				}
			},
		);
	}

	moveTile(x: number, y: number) {
		const emptyTile = this.findEmptyTile();
		this.tiles[emptyTile.x][emptyTile.y] = this.tiles[x][y];
		this.tiles[x][y] = null;
		this.tweens.add({
			targets: this.tiles[emptyTile.x][emptyTile.y],
			x: emptyTile.x * 200 + this.gameWidth / 2 - 200,
			y: emptyTile.y * 200 + this.gameHeight / 2 - 200,
			duration: 100,
		});
		this.checkWin();
	}

	canMoveTile(x: number, y: number) {
		const emptyTile = this.findEmptyTile();
		return (
			(x === emptyTile.x && Math.abs(y - emptyTile.y) === 1) ||
			(y === emptyTile.y && Math.abs(x - emptyTile.x) === 1)
		);
	}

	findEmptyTile() {
		for (let i = 0; i < this.tiles.length; i++) {
			for (let j = 0; j < this.tiles[i].length; j++) {
				if (this.tiles[i][j] == null) {
					return { x: i, y: j };
				}
			}
		}
		return { x: -1, y: -1 };
	}

	checkWin() {
		if (this.isShuffling) {
			return;
		}
		for (let i = 0; i < this.tiles.length; i++) {
			for (let j = 0; j < this.tiles[i].length; j++) {
				if (this.tiles[i][j] !== this.correctTiles[i][j]) {
					return;
				}
			}
		}
		this.clearCount++;
		this.scene.start('gameScene', {
			clearCount: this.clearCount,
			timer: 60 - this.clearCount,
		});
	}

	shuffleTiles() {
		this.isShuffling = true;
		const directions = [
			{ x: -1, y: 0 },
			{ x: 1, y: 0 },
			{ x: 0, y: -1 },
			{ x: 0, y: 1 },
		];

		for (let i = 0; i < 1000; i++) {
			const emptyTile = this.findEmptyTile();
			const validDirections = directions.filter((direction) =>
				this.isValidTile(emptyTile.x + direction.x, emptyTile.y + direction.y),
			);
			if (validDirections.length === 0) {
				break;
			}
			const direction = Phaser.Utils.Array.GetRandom(validDirections);
			this.moveTile(emptyTile.x + direction.x, emptyTile.y + direction.y);
		}
		this.isShuffling = false;
	}

	isValidTile(x: number, y: number) {
		return x >= 0 && x < 3 && y >= 0 && y < 3;
	}

	update(time: number, delta: number) {}
}

const config: Phaser.Types.Core.GameConfig = {
	type: Phaser.AUTO,
	title: '8パズル',
	width: GAME_CONFIG.DEFAULT_WIDTH,
	height: GAME_CONFIG.DEFAULT_HEIGHT,
	scene: [TitleScene, GameScene],
	scale: {
		parent: 'phaser-game',
		mode: Phaser.Scale.FIT,
		fullscreenTarget: 'phaser-game',
		autoCenter: Phaser.Scale.CENTER_HORIZONTALLY,
		zoom: Phaser.Scale.MAX_ZOOM,
	},
};

export default function initPhaserGame(parent: HTMLElement) {
	if (config.scale) {
		config.scale.parent = parent;
		config.scale.fullscreenTarget = parent;
	}
	return new Phaser.Game(config);
}
