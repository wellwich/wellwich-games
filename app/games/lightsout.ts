import Phaser from 'phaser';

const D_WIDTH = 640;
const D_HEIGHT = 480;

const cellSize = 128;

class TitleScene extends Phaser.Scene {
	constructor() {
		super('titleScene');
	}
	create() {
		const background = this.add.rectangle(
			0,
			0,
			D_WIDTH,
			D_HEIGHT,
			0x444444,
			0.5,
		);
		background.setOrigin(0, 0);
		background.setInteractive();
		const fullScreen = this.add
			.text(D_WIDTH - 16, 16, 'Full Screen', {
				fontSize: '16px',
				fontFamily: '"font"',
			})
			.setOrigin(1, 0);
		fullScreen.setInteractive();
		fullScreen.on('pointerup', () => {
			if (this.scale.isFullscreen) {
				this.scale.stopFullscreen();
			} else {
				this.scale.startFullscreen();
			}
		});
		const title = this.add
			.text(D_WIDTH / 2, D_HEIGHT / 2 - 50, 'ライツアウト', {
				fontSize: '64px',
				color: '#fff',
				fontFamily: 'monospace',
			})
			.setOrigin(0.5, 0.5);
		const start = this.add
			.text(D_WIDTH / 2, D_HEIGHT / 2 + 100, 'Tap to start', {
				fontSize: '16px',
				fontFamily: '"font"',
			})
			.setOrigin(0.5, 0.5);
		// 画面のどこかをタップしたらゲーム画面へ
		background.on('pointerup', () => {
			this.scene.start('gameScene', { limit: 30 });
		});
		if (window.matchMedia?.('(max-device-width: 640px)').matches) {
			const background = this.add.rectangle(0, 0, D_WIDTH, D_HEIGHT, 0x444444);
			background.setInteractive();
			background.setOrigin(0, 0);
			this.add
				.text(D_WIDTH / 2, D_HEIGHT / 2, '横画面にしてください', {
					fontSize: '64px',
					color: '#fff',
					fontFamily: 'monospace',
				})
				.setFontSize(32)
				.setOrigin(0.5);
			const fixScreen = this.add
				.text(D_WIDTH / 2, D_HEIGHT / 2 + 150, '横画面にしたらここをタップ', {
					fontSize: '64px',
					color: '#fff',
					fontFamily: 'monospace',
				})
				.setFontSize(16)
				.setOrigin(0.5);
			fixScreen.setInteractive();
			fixScreen.on('pointerup', () => {
				this.scale.startFullscreen();
				this.scene.start('titleScene');
			});
		}
	}
}

class GameScene extends Phaser.Scene {
	constructor() {
		super('gameScene');
	}

	private limit!: number;

	init(data: { limit: number }) {
		this.limit = data.limit;
	}

	preload() {
		this.load.image('gauge', 'https://assets.wellwich.com/lightsout/gauge.png');
	}

	create() {
		const background = this.add
			.rectangle(0, 0, D_WIDTH, D_HEIGHT, 0x444444, 0.5)
			.setOrigin(0, 0);
		const timer = this.time.addEvent({
			delay: 500,
			callback: () => {
				gaugeMask.height += Math.ceil(224 / this.limit);
				if (gaugeMask.height >= 160) {
					gauge.setFillStyle(0xff0000);
				}
				if (gaugeMask.height >= 224) {
					timer.remove();
					const BG = this.add
						.rectangle(0, 0, D_WIDTH, D_HEIGHT, 0x000000, 0.5)
						.setOrigin(0, 0);
					BG.setInteractive();
					const clearText = this.add.text(
						D_WIDTH / 2,
						D_HEIGHT / 2,
						'GAME OVER',
						{
							fontSize: '64px',
							color: '#ffffff',
						},
					);
					clearText.setOrigin(0.5, 0.5);
					// クリックされたらリロード
					BG.on('pointerup', () => {
						this.scene.start('titleScene');
					});
				}
			},
			repeat: this.limit - 1,
		});
		const gaugeX = 64;
		const gaugeY = 96;
		const levelText = this.add
			.text(
				gaugeX - 16,
				32,
				`Lv.${(31 - this.limit).toString().padStart(2, '0')}`,
				{ fontSize: '32px', color: '#fff', fontFamily: 'monospace' },
			)
			.setOrigin(0, 0);
		const gauge = this.add
			.rectangle(gaugeX + 16, gaugeY + 16, 32, 224, 0xffff00)
			.setOrigin(0, 0);
		const gaugeMask = this.add
			.rectangle(gaugeX + 16, gaugeY + 16, 32, 0, 0x000000)
			.setOrigin(0, 0);
		const guageframe = this.add.image(gaugeX, gaugeY, 'gauge').setOrigin(0, 0);
		// ライツアウトの盤面を作成
		const board = [
			[0, 0, 0],
			[0, 0, 0],
			[0, 0, 0],
		];
		// boardの値を0か1にする
		board.forEach((row, y) => {
			row.forEach((cell, x) => {
				board[y][x] = Phaser.Math.Between(0, 1);
			});
		});
		// ライツアウトの盤面を描画
		const whiteCell = this.add.graphics();
		whiteCell.fillStyle(0xffffff);
		whiteCell.fillRect(0, 0, cellSize, cellSize);
		whiteCell.generateTexture('whiteCell', cellSize, cellSize);
		whiteCell.destroy();
		const blackCell = this.add.graphics();
		blackCell.fillStyle(0x000000);
		blackCell.fillRect(0, 0, cellSize, cellSize);
		blackCell.generateTexture('blackCell', cellSize, cellSize);
		blackCell.destroy();

		for (const [y, row] of board.entries()) {
			for (const [x, cell] of row.entries()) {
				const cellSprite = this.add.sprite(
					0,
					0,
					cell === 0 ? 'whiteCell' : 'blackCell',
				);
				cellSprite.setOrigin(0, 0);
				cellSprite.setPosition(x * cellSize + 192, y * cellSize + 48);
				cellSprite.setInteractive();
				cellSprite.on('pointerdown', (pointer: Phaser.Input.Pointer) => {
					// クリックされたセルを反転
					cellSprite.setTexture(
						cellSprite.texture.key === 'whiteCell' ? 'blackCell' : 'whiteCell',
					);
					// 上下左右のセルを反転
					const neighbors = [
						{ x: x - 1, y },
						{ x: x + 1, y },
						{ x, y: y - 1 },
						{ x, y: y + 1 },
					];
					for (const neighbor of neighbors) {
						if (
							neighbor.x >= 0 &&
							neighbor.x < board[0].length &&
							neighbor.y >= 0 &&
							neighbor.y < board.length
						) {
							const neighborSprite = this.children.getByName(
								`${neighbor.x}-${neighbor.y}`,
							) as Phaser.GameObjects.Sprite;
							neighborSprite.setTexture(
								neighborSprite.texture.key === 'whiteCell'
									? 'blackCell'
									: 'whiteCell',
							);
						}
					}
					// クリア判定
					const isClear = board.every((row, y) => {
						return row.every((cell, x) => {
							const cellSprite = this.children.getByName(
								`${x}-${y}`,
							) as Phaser.GameObjects.Sprite;
							return cellSprite.texture.key === 'whiteCell';
						});
					});
					if (isClear) {
						const BG = this.add
							.rectangle(0, 0, D_WIDTH, D_HEIGHT, 0x000000, 0.5)
							.setOrigin(0, 0);
						BG.setInteractive();
						const clearText = this.add.text(
							D_WIDTH / 2,
							D_HEIGHT / 2,
							'CLEAR',
							{
								fontSize: '64px',
								color: '#ffffff',
							},
						);
						clearText.setOrigin(0.5, 0.5);
						// クリックされたらリロード
						const delay = this.time.addEvent({
							delay: 1000,
							callback: () => {
								this.scene.restart({ limit: this.limit - 1 });
							},
						});
					}
				});
				cellSprite.setName(`${x}-${y}`);
			}
		}
	}
}

const config: Phaser.Types.Core.GameConfig = {
	type: Phaser.AUTO,
	title: 'ライツアウト',
	width: D_WIDTH,
	height: D_HEIGHT,
	scene: [TitleScene, GameScene],
	scale: {
		parent: 'phaser-game',
		mode: Phaser.Scale.NONE,
		fullscreenTarget: 'phaser-game',
		autoCenter: Phaser.Scale.CENTER_HORIZONTALLY,
		zoom: Phaser.Scale.MAX_ZOOM,
	},
};

export default function initPhaserGame(parent: HTMLElement) {
	if (config.scale) {
		config.scale.parent = parent;
		config.scale.fullscreenTarget = parent;
	}
	return new Phaser.Game(config);
}
