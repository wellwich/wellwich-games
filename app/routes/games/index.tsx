import { createRoute } from 'honox/factory';
import Header from '../../components/Header';
import GameList from '../../islands/GameList/GameList';

export default createRoute((c) => {
	const name = c.req.param('name');
	return c.render(
		<div>
			<Header />
			<h2 class="text-4xl font-bold">{name}</h2>
			<div class="max-w-lg mx-auto">
				<GameList />
			</div>
		</div>,
		{ title: name },
	);
});
