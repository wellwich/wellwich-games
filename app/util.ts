function waitForElement(
	container: HTMLElement,
	selector: string,
): Promise<void> {
	return new Promise<void>((resolve) => {
		const observer = new MutationObserver((_mutations, obs) => {
			const element = container.querySelector(selector);
			if (element) {
				obs.disconnect();
				resolve();
			}
		});
		observer.observe(container, { childList: true, subtree: true });
	});
}

export { waitForElement };
