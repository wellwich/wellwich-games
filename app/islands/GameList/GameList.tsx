import { useEffect, useState } from 'hono/jsx';
import games from '../../game-info';

const GameList = () => {
	const [gameList, setGameList] = useState<
		{ name: string; path: string; description: string; image: string }[]
	>([]);

	useEffect(() => {
		setGameList(games);
	}, []);

	return (
		<ul class="grid grid-cols-2 gap-4">
			{gameList.map((game) => (
				<li key={game.name} class="m-4 hover:bg-gray-200">
					<a href={`/games/${game.path}`}>
						<div class="p-4 border-solid border-2 border-gray-400 h-64">
							<img
								src={game.image}
								alt={game.name}
								class="w-full h-32 object-cover border-solid border-2 border-gray-400"
							/>
							<h4 class="text-xl font-bold">{game.name}</h4>
							<p class="overflow-hidden text-ellipsis line-clamp-3">
								{game.description}
							</p>
						</div>
					</a>
				</li>
			))}
		</ul>
	);
};

export default GameList;
