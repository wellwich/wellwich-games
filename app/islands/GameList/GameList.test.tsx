import { render } from 'hono/jsx/dom';
import games from '../../game-info';
import { waitForElement } from '../../util';
import GameList from './GameList';

describe('ゲームリストコンポーネント', () => {
	it('レンダリングされるべき', async () => {
		const container = document.createElement('div');
		render(<GameList />, container);

		await waitForElement(container, 'ul');

		expect(container).toMatchSnapshot();
	});

	it('ゲームリストが正常に表示されるべき', async () => {
		const container = document.createElement('div');
		render(<GameList />, container);

		await waitForElement(container, 'li');

		const gameList = container.querySelectorAll('li');
		expect(gameList.length).toBe(games.length);
	});
});
