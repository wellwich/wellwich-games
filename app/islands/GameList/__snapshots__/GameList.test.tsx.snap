// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ゲームリストコンポーネント レンダリングされるべき 1`] = `
<div>
  <ul
    class="grid grid-cols-2 gap-4"
  >
    <li
      class="m-4 hover:bg-gray-200"
    >
      <a
        href="/games/color-tiles"
      >
        <div
          class="p-4 border-solid border-2 border-gray-400 h-64"
        >
          <img
            alt="カラータイル"
            class="w-full h-32 object-cover border-solid border-2 border-gray-400"
            src="https://assets.wellwich.com/web/color-tiles_thumb.webp"
          />
          <h4
            class="text-xl font-bold"
          >
            カラータイル
          </h4>
          <p
            class="overflow-hidden text-ellipsis line-clamp-3"
          >
            色のついたタイルをクリックして、同じ色のタイルを消していこう。
          </p>
        </div>
      </a>
    </li>
    <li
      class="m-4 hover:bg-gray-200"
    >
      <a
        href="/games/eight-puzzle"
      >
        <div
          class="p-4 border-solid border-2 border-gray-400 h-64"
        >
          <img
            alt="8パズル"
            class="w-full h-32 object-cover border-solid border-2 border-gray-400"
            src="https://assets.wellwich.com/web/eight-puzzle_thumb.webp"
          />
          <h4
            class="text-xl font-bold"
          >
            8パズル
          </h4>
          <p
            class="overflow-hidden text-ellipsis line-clamp-3"
          >
            数字を並べ替えて、元の順番に戻そう。
          </p>
        </div>
      </a>
    </li>
    <li
      class="m-4 hover:bg-gray-200"
    >
      <a
        href="/games/lightsout"
      >
        <div
          class="p-4 border-solid border-2 border-gray-400 h-64"
        >
          <img
            alt="ライツアウト"
            class="w-full h-32 object-cover border-solid border-2 border-gray-400"
            src="https://assets.wellwich.com/web/lightsout_thumb.webp"
          />
          <h4
            class="text-xl font-bold"
          >
            ライツアウト
          </h4>
          <p
            class="overflow-hidden text-ellipsis line-clamp-3"
          >
            すべてのライトをつけよう。
          </p>
        </div>
      </a>
    </li>
    <li
      class="m-4 hover:bg-gray-200"
    >
      <a
        href="/games/quiz"
      >
        <div
          class="p-4 border-solid border-2 border-gray-400 h-64"
        >
          <img
            alt="クイズ"
            class="w-full h-32 object-cover border-solid border-2 border-gray-400"
            src="https://assets.wellwich.com/web/quiz_thumb.webp"
          />
          <h4
            class="text-xl font-bold"
          >
            クイズ
          </h4>
          <p
            class="overflow-hidden text-ellipsis line-clamp-3"
          >
            クイズに答えて、正解数を競おう。
          </p>
        </div>
      </a>
    </li>
  </ul>
</div>
`;
